README.md
setPlaceholders
===============

A simple MODX 2.x snippet for getting fields and setting placeholders.

Placeholders have a number of useful applications.  Essentially they can be used as local variables in a template or chunk.  They&rsquo;ve always been simple to set from within a snippet; setPlaceholders makes it easy to set them without any PHP coding.

Additionally setPlaceholders provides a convenient way of gathering up various data (fields and TVs from a resource and any of its ancestors, arbitrary and fallback values, even $\_GET and $\_POST variables) and putting them into placeholders or returning them as output.  Most of this is achieved through a handy syntax which allows fields names to be prefixed. For instance parent.tv.myTV will get the value of myTV from the selected resource&rsquo;s parent.

setPlaceholders may also be used as an output filter to store the result of an expression or snippet call for use, say, farther down in the template.

Let&rsquo;s start with some examples.

Examples
--------


* Showing most of the functionality:<br>
```[[setPlaceholders? &id=`13` ```<br>
```&fields=`pagetitle || tv.someTV !! [[snippetX]] || get.person`  ```<br>
```&placeholders=`color==#a35c0a || params == w=240&h=360&q=65` ```<br>
```]]```<br>
Sets the placeholders:<br>
```[[+sph.pagetitle]]``` Resource 13's pagetitle<br>
```[[+sph.tv.someTV]]``` value of resource 13's TV someTV, or the output from snippetX if not found<br>
```[[+sph.get.person]]``` value of $_GET['person']<br>
```[[+color]]``` #a35c0a<br>
```[[+params]]``` w=240&h=360&q=65

* As a getResourceField replacement:<br>
```[[setPlaceholders? &fields=`parent.tv.someTV !! No such TV` &output=`1`]]```<br>
Returns the value of someTV for the current resource's parent, or &ldquo;No such TV&rdquo; if it&rsquo;s empty or not found.  It also puts this value in ```[[+sph.parent.tv.someTV]]```<br>
This is the equivalent of:<br>
```[[getResourceField? &id=`[[*parent]]` &field=`someTV` &isTV=`1` &processTV=`1` &default=`No such TV`]]```

* As an output filter:<br>
``` [[*someTV:eq=`foo`:then=`bar`:else=`nobar`:setPlaceholders=`fb`]] ```<br>
Returns the value of the expression and also stores it in [[+fb]] for later use.

* Using &amp;prefix to set placeholders for the same field from different resources:<br>
```[[setPlaceholders? &id=`13` &field=`tv.varA` &prefix=`r13.`]]```<br>
```[[setPlaceholders? &id=`14` &field=`tv.varA` &prefix=`r14.`]]```<br>
Sets the placeholders:<br>
```[[+r13.tv.varA]]``` and ```[[+r14.tv.varA]]```

* One more example, with some lesser-used options:<br>
```[[setPlaceholders? &fields=`parent.pagetitle||parent.parent.pagetitle` &prefix=`` &output=`1`
&delimiter=` > `]]```<br>
Returns: Page Title #1 > Page Title #2 (of course it&rsquo;d be the actual pagetitles)<br>
Sets the placeholders:<br>
```[[+parent.pagetitle]]``` The current resource&rsquo;s parent&rsquo;s pagetitle<br>
```[[+parent.parent.pagetitle]]``` The current resource&rsquo;s grandparent&rsquo;s pagetitle


Properties
----------

*When called as a snippet:*

<table>
<tr><th>Name</th><th>Description</th><th>Default</th></tr>
<tr>
  <td>&amp;id</td>
  <td>The resource id to use</td>
  <td>current resource</td>
</tr><tr>
  <td>&amp;fields</td>
  <td>A list of fields to retrieve, separated by ||<br>
    <em>Field names may be prefixed by</em>:<br>
    <strong>tv.</strong> – to get a TV<br><strong>parent.</strong> – to get a field from the resource&rsquo;s parent. Can be used multiple times to get a parent&rsquo;s parent, etc. and can be a tv (ex: parent.parent.tv.someTV).<br>
    <strong>get.</strong> – to get a variable from $_GET<br>
    <strong>post.</strong> – to get a variable from $_POST<br><br>
    Specify a default value by appending !! value to the field name.  This will be used if the field is empty or not found. Ex: longtitle !! Default Title<br>
    Placeholder names are a prefix (see &amp;prefix) plus the field name. Ex: &amp;fields=`tv.someTV || parent.id` will set the placeholders [[+sph.tv.someTV]] and [[+sph.parent.id]]</td>
  <td></td>
</tr><tr>
  <td>&amp;prefix</td>
  <td>Prepended to all placeholder names from &amp;fields to reduce the likelihood of placeholder name conflicts. Not added to any user-specified placeholders in &amp;placeholders. Can be set to an empty string to eliminate the prefix.</td>
  <td>sph.</td>
</tr><tr>
  <td>&amp;placeholders</td>
  <td>An optional list of user-defined placeholders to set<br>Format: name1==value1 || name2==value2</td>
  <td></td>
</tr><tr>
  <td>&amp;output</td><td>Output mode.<br><em>No</em> (0): only set placeholders<br><em>Yes</em> (1): also output the value of any placeholders. This allows the snippet to be used like getResourceField. Multiple values will be separated by &amp;delimiter.</td>
  <td>No</td>
</tr><tr>
  <td>&amp;delimiter</td>
  <td>Separates values when results are returned as a string (i.e. &amp;output=`1`)</td>
  <td>,</td>
</tr>
</table>

<br>

*When used as an output filter:*

* setPlaceholders takes one argument: the name of the placeholder to set.

Notes
-----

* It's fine to add white space around ||, ==, and !! for legibility. It'll get trimmed off.
* If you&rsquo;re using get or post in &amp;fields, you&rsquo;ll want to call setPlaceholders uncached. And since these variables can contain anything–including various hack attempts–be sure to validate and sanitize them properly before passing them on as input to something else, especially a SQL query.

changelog.txt
Changelog for setPlaceholders

setPlaceholders 1.0.0-pl
---------------------------------
Initial Version

core/components/setplaceholders/elements/snippets/setplaceholders.snippet.php
<?php
/**
 * setPlaceholders
 * Copyright 2012 Jason Grant
 *
 * setPlaceholders is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * setPlaceholders is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * setPlaceholders; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *
 * @package setPlaceholders
 * @author Jason Grant
 * @version 1.0.0-pl
 */

/**
 * Full documentation, bug reports, etc.
 * https://github.com/oo12/setPlaceholders
 *
 * Variables
 * ---------
 *
 * @var modX $modx
 * @var input $input
 * @var options $options
 *
 *
 * Properties
 * ----------
 *
 * @property id - (integer)
 * @property fields - (string)
 * @property prefix - (string)
 * @property placeholders - (string)
 * @property output - (boolean)
 * @property delimiter - (string)
 *
 * See the default properties for a description of each.
 * Parameters are all optional, but you'll want to specify either &fields
 * or &placeholders or both to accomplish anything.
 *
 *
 * Examples
 * --------
 *
 * [[setPlaceholders? &id=`13` &fields=`pagetitle || tv.someTV || get.person`
 *   &placeholders=`color==#a35c0a || params == w=240&h=360&q=65`]]
 *
 * Sets the placeholders:
 *   [[+sph.pagetitle]] - Resource 13's pagetitle
 *   [[+sph.tv.someTV]] - value of resource 13's TV someTV
 *   [[+sph.get.person]] - value of $_GET['person'] (from the URL)
 *   [[+color]] - #a35c0a
 *   [[+params]] - w=240&h=360&q=65
 *
 *
 * As a getResourceField replacement:
 *
 * [[setPlaceholders? &fields=`parent.tv.someTV !! No such TV` &output=`1`]]
 *
 * Returns the value of someTV for the current resource's parent, or
 * "No such TV" if it's not found.  It also puts this value in
 * [[+sph.parent.tv.someTV]]
 *
 *
 * As an output filter:
 *
 * [[*someTV:eq=`foo`:then=`bar`:else=`nobar`:setPlaceholders=`fb`]]
 *
 * Returns the value of the expression and also stores it in [[+fb]]
 *
 *
 * Whitespace around || and !! and == is fine; it'll be trimmed off if present.
 *
 * @package setPlaceholders
 **/

if ( isset($input) && $options ) {  // if we're being used as an output filter
	$modx->setPlaceholder($options, $input);
	return $input;
}

// handle options
$id = $id ? intval($id) : $modx->resource->get('id');
$fields = $fields ? explode('||', $fields) : false;
$prefix = isset($prefix) ? $prefix : 'sph.';
$placeholders = $placeholders ? explode('||', $placeholders) : NULL;
$delimiter = isset($delimiter) ? $delimiter : ',';

if ($fields) {  // if we'll need to be looking up anything..
	$resource = $modx->getObject('modResource', $id);
}

$p = array();  // placeholder storage
$parents = array(); // cache for parent resources

foreach ($fields as $field) {
	// prep $field, separate out any default and prefixes
	$field = explode('!!', $field);
	$fieldName = $field[0] = trim( $field[0] );
	$fieldPrefixes = explode('.', $fieldName);

	$doc = $resource;
	$value = NULL;

	if ( $fieldPrefixes[0] === 'get' ) {
		$value = $_GET[substr($fieldName, 4)];
	}
	elseif ( $fieldPrefixes[0] === 'post' ) {
		$value = $_POST[substr($fieldName, 5)];
	}
	else {
		for ($idx = 0; $fieldPrefixes[$idx] === 'parent' && $doc; ++$idx) {
			if ( !$parents[$idx] )  {
				$parents[] = $modx->getObject('modResource', $doc->get('parent'));  // Get the parent object and cache it
			}
			$doc = $parents[$idx];
			$fieldName = substr($fieldName, 7);
		}
		if ($doc) {  // if we've got a valid resource
			if ( $fieldPrefixes[$idx] === 'tv' ) {
				$value = $doc->getTVValue( substr($fieldName, 3) );
			}
			else {
				$value = $doc->get($fieldName);
			}
		}
	}

	if ( $value == '' && $field[1] ) {  // if we didn't find a value, use the default
		$value = trim( $field[1] );
	}
	if ( $value != '') {
		$p[ $prefix . $field[0] ] = $value;  // key: prefix + field name
	}
}

foreach ($placeholders as $placeholder) { // add any user-defined placeholders
	$ph = explode('==', $placeholder);
	$p[ trim($ph[0]) ] = trim($ph[1]);
}

// Output our results
$modx->setPlaceholders($p);
return ($output ? implode($delimiter, $p) : '');